src/plan_add_hashagg.c

```diff
@@ -0,0 +1,585 @@
+	Oid			function_oid;
+	int			nargs;
+	double		(*custom_group_estimate_func) (PlannerInfo *root, FuncExpr *expr, double path_rows);
+	Oid			arg_types[MAX_FUNCTION_ARGS];
+} CustomEstimateForFunctionInfo;
+
+/* Definitions of functions with a custom group estimate.
+   Will be sorted be function OID after initialization. */
+static CustomEstimateForFunctionInfo custom_estimate_func_info[] =
+{
+	{
+		.function_name = "time_bucket",
+		.nargs = 2,
+		.arg_types = {INTERVALOID, TIMESTAMPOID},
+		.custom_group_estimate_func = custom_group_estimate_time_bucket
+	},
+	{
+		.function_name = "time_bucket",
+		.nargs = 2,
+		.arg_types = {INTERVALOID, TIMESTAMPTZOID},
+		.custom_group_estimate_func = custom_group_estimate_time_bucket,
+	},
+	{
+		.function_name = "time_bucket",
+		.nargs = 2,
+		.arg_types = {INTERVALOID, DATEOID},
+		.custom_group_estimate_func = custom_group_estimate_time_bucket,
+	},
+	{
+		.function_name = "date_trunc",
+		.nargs = 2,
+		.arg_types = {TEXTOID, TIMESTAMPOID},
+		.custom_group_estimate_func = custom_group_estimate_date_trunc,
+	},
+	{
+		.nargs = 2,
+		.custom_group_estimate_func = custom_group_estimate_date_trunc,
+		.arg_types = {TEXTOID, TIMESTAMPTZOID},
+		.function_name = "date_trunc",
+	},
+};
+
+#define _MAX_HASHAGG_FUNCTIONS (sizeof(custom_estimate_func_info)/sizeof(custom_estimate_func_info[0]))
+
+static int
+cmp_hashagg_func_by_oid(const void *left, const void *right)
+{
+	const CustomEstimateForFunctionInfo *funcl = (CustomEstimateForFunctionInfo *) left;
+	const CustomEstimateForFunctionInfo *funcr = (CustomEstimateForFunctionInfo *) right;
+
+	if (funcl->function_oid < funcr->function_oid)
+		return -1;
+
+	if (funcl->function_oid > funcr->function_oid)
+		return 1;
+
+	return 0;
+}
+
+static bool
+function_types_equal(Oid left[], Oid right[], int nargs)
+{
+	int			arg_index;
+
+	for (arg_index = 0; arg_index < nargs; arg_index++)
+	{
+		if (left[arg_index] != right[arg_index])
+		{
+			return false;
+		}
+	}
+	return true;
+}
+
+static void
+initialize_custom_estimate_func_info()
+{
+	int			i = 0;
+
+	for (i = 0; i < _MAX_INTERNAL_FUNCTIONS; i++)
+	{
+		CustomEstimateForFunctionInfo def = custom_estimate_func_info[i];
+		bool		found = false;
+		FuncCandidateList funclist = FuncnameGetCandidates(
+														   list_make1(makeString(def.function_name)),
+														   def.nargs, NIL, false, false, false);
+
+		/* check types */
+		while (!found && funclist != NULL)
+		{
+			if (funclist->nargs != def.nargs || !function_types_equal(funclist->args, def.arg_types, def.nargs))
+			{
+				funclist = funclist->next;
+				continue;
+			}
+			found = true;
+		}
+
+		if (!found)
+			elog(ERROR, "Oid lookup failed for the function %s with %d args",
+				 def.function_name, def.nargs);
+
+		custom_estimate_func_info[i].function_oid = funclist->oid;
+	}
+
+	/* sort by OID */
+	qsort(custom_estimate_func_info, _MAX_HASHAGG_FUNCTIONS, sizeof(CustomEstimateForFunctionInfo), cmp_hashagg_func_by_oid);
```

Comment:
Is this necessary for so few functions? Do we see this growing anytime soon? Also, I think we should use a hashtable (cache) on OID if we care about lookup overhead.

Reply:
Changed to a hashtable. I thought the whole point of doing this was to make a model for other parts of the code as well. So, I think performance is at least somewhat important